include/boost/math/tools/polynomial.hpp

```diff
@@ -558,59 +562,52 @@ class polynomial :
 
 
 template <class T>
-inline polynomial<T> operator + (const polynomial<T>& a, const polynomial<T>& b)
+inline polynomial<T> operator + (polynomial<T> a, const polynomial<T>& b)
 {
-   polynomial<T> result(a);
-   result += b;
-   return result;
+   a += b;
+   return a;
 }
 
 template <class T>
-inline polynomial<T> operator - (const polynomial<T>& a, const polynomial<T>& b)
+inline polynomial<T> operator - (polynomial<T> a, const polynomial<T>& b)
 {
-   polynomial<T> result(a);
-   result -= b;
-   return result;
+   a -= b;
+   return a;
 }
 
 template <class T>
-inline polynomial<T> operator * (const polynomial<T>& a, const polynomial<T>& b)
+inline polynomial<T> operator * (polynomial<T> a, const polynomial<T>& b)
```

Comment:
I think we need to be careful about changing the parameter type to an object type rather than an lvalue-ref.

We have 4 possibilities here:

2 l-value refs: need a new temporary for the result.
left r-value ref, right l-value ref: reuse left argument for the result.
right r-value ref, left lvalue ref: reuse right argument for the result (except division!)
both rvalue-refs: reuse either for the result.

Your change handles the first two but not the last two, and in order to avoid overload ambiguities, I *think* the only way to do this right is to provide all 4 overloads.  I realise that sucks, but it does work well (or at least it did for Boost.Multiprecision when I did this there).

For multiplication, I think we should also rewrite the *= operator to carry out 2-arg multiplication since there isn't really an efficient in-place multiplication, something like:

`
void multiply(const Polynomial& a, const polynomial& b)
{
  // allocates temporary vector for result, then swaps it into *this.
}
`

The code would be much the same as now, but I think we can avoid at least one copy doing this when we're doing 2 arg multiplication and both in-place  and 2 arg multiplication can simply call the above routine?